---
title: "16s boxplot"
author: "Caroline Wheeler"
date: "3/2/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(vegan)
library(reshape2)
```

read in data 
```{r}
samp_match <- read.csv("../data/16s_sample_matching.csv")
krakenmet = read.delim("/fs/ess/PAS1695/exoticpipe/external-data/kraken2-metaphlan-noplants.txt",
                         header = F, stringsAsFactors = F) %>%
    dplyr::rename("Taxonomy" = "V1")
exogieo <- read.csv("../data/16s_counts_long.csv") %>%
  dplyr::rename("microbe" = name,
         "counts" = new_est_reads)
# # read in 16s counts wide
# exogieo <- read.csv("../data/16s_taxa_allLevels_counts_wide.csv") %>%
#   # filter out human reads
#   filter(microbe != "Homo.sapiens")
exotic <- readRDS("/fs/ess/PAS1695/projects/exotic/data/drake-output/2022-03-08/tcc.counts.RDS")
```

crop 16s sample labels
```{r}
samp_match <- samp_match %>%
  mutate(crop16s = substr(samp16s, 9, length(samp16s)))
```

exotic <- wide to long
```{r}
exotic <- gather(exotic, key="microbe", value="counts", 2:ncol(exotic))
```

```{r}
taxkey <- krakenmet %>%
  dplyr::filter(grepl("s__", Taxonomy)) %>%
  mutate(microbe = gsub(".*s__(.*)", "\\1", Taxonomy),
          microbe = make.names(microbe))
```

combine with taxkey
```{r}
exogieo.all <- merge(exogieo, taxkey)
exotic.all <- merge(exotic, taxkey)
```

function separate taxa into levels 
```{r}
get_levels <- function(exora){
  exora <- exora %>%
   dplyr::select(sample, microbe, Taxonomy, counts) %>%
    mutate(domain = gsub("(d__\\w+).*", "\\1", Taxonomy),
           kingdom = ifelse(grepl("k__", Taxonomy),gsub(".*(k__\\w+).*", "\\1", Taxonomy), NA),
           phylum = ifelse(grepl("p__", Taxonomy),gsub(".*(p__\\w+).*", "\\1", Taxonomy), NA),
           order = ifelse(grepl("o__", Taxonomy),gsub(".*(o__\\w+).*", "\\1", Taxonomy), NA),
           class = ifelse(grepl("c__", Taxonomy),gsub(".*(c__\\w+).*", "\\1", Taxonomy), NA),
           family = ifelse(grepl("f__", Taxonomy),gsub(".*(f__\\w+).*", "\\1", Taxonomy), NA),
           genus = ifelse(grepl("g__", Taxonomy),gsub(".*(g__\\w+).*", "\\1", Taxonomy), NA),
           species = gsub(".*(s__.*)", "\\1", Taxonomy)
    ) %>%
    mutate(genus = ifelse(is.na(genus), paste0("g__unclassified-", species), genus),
           family = ifelse(is.na(family), paste0("f__unclassified-", gsub("g__unclassified-", "", genus)), family),
           order = ifelse(is.na(order), paste0("o__unclassified-", gsub("f__unclassified-", "", family)), order),
           class = ifelse(is.na(class), paste0("c__unclassified-", gsub("o__unclassified-", "", order)), class),
           phylum = ifelse(is.na(phylum), paste0("p__unclassified-", gsub("c__unclassified-", "", class)),phylum),
           kingdom = ifelse(is.na(kingdom), paste0("k__unclassified-", gsub("p__unclassified-", "", phylum)), kingdom),
           domain = ifelse(is.na(domain), paste0("d__unclassified-", gsub("k__unclassified-", "", kingdom)), domain)) %>%
    dplyr::select(sample, microbe, Taxonomy, domain, kingdom, phylum, class, order, family, genus, species, counts)
  return(exora)
}
```


```{r}
exotic.all <- get_levels(exotic.all)
exogieo.all <- get_levels(exogieo.all)
```

filter exotic to only bacteria 
```{r}
exotic.all <- exotic.all %>%
  filter(domain == "d__Bacteria")
```

get prevalence 1 if present, 0 otherwise
```{r}
exogieo.prev <- exogieo.all %>%
  select(sample, genus, counts) %>% 
  group_by(sample, genus) %>% 
  summarize(count = sum(counts)) %>%
  mutate(prev = ifelse(count > 0, 1, 0)) %>%
  select(-count)

exotic.prev <- exotic.all %>%
  select(sample, genus, counts) %>% 
  group_by(sample, genus) %>% 
  summarize(count = sum(counts)) %>%
  mutate(prev = ifelse(count > 0, 1, 0)) %>%
  select(-count)
```


filter exotic to SLids in paired data
```{r}
# exotic.ra <- exotic.ra %>%
#   filter(sample %in% samp_match$RNAseq)
exotic.prev <- exotic.prev %>%
  filter(sample %in% samp_match$RNAseq)
```

combine the 16s and RNAseq into one df and spread
```{r}
# both <- rbind(exotic.ra, exogieo.ra) %>%
#   # drop count
#   select(-count) %>%
#   spread(key = genus, value = rel.abun, fill = 0)
exogieo.prev$sample <- as.character(exogieo.prev$sample)
both <- rbind(exotic.prev, exogieo.prev) %>%
  spread(key = genus, value = prev, fill = 0)
```

make genus row names and remove genus col
```{r}
rnames <- both$sample
both <- both[, -1]
rownames(both) <- rnames
both <- as.matrix(both)
```

Bray Curtis distance calculation
```{r}
set.seed(19971030)
dist <- vegdist(both, method = "bray")
# binary = FALSE, upper = TRUE, na.rm = TRUE
```

distance matrix to long
```{r}
dist.res <- melt(as.matrix(dist), varnames = c("RNAseq", "samp16s"))
```

filter columns to the correct sample type
```{r}
dist.res <- dist.res %>%
  filter(RNAseq %in% samp_match$RNAseq,
         samp16s %in% samp_match$crop16s)
```

prep pt id, paired, and specimen type cols
```{r}
pt.rna <- samp_match %>%
  select(patient.id, RNAseq) %>%
  dplyr::rename("pt.id.rna" = patient.id) %>%
  filter(!is.na(RNAseq)) %>%
  distinct()

pt.16s <- samp_match %>%
  select(patient.id, crop16s, specimen.type) %>%
  dplyr::rename("pt.id.16s" = patient.id,
         "samp16s" = crop16s)
```

add pt id, paired, and speciment type cols
```{r}
tmp <- merge(dist.res, pt.rna)
dist.res <- merge(tmp, pt.16s)
```

get paired column
```{r}
dist.res <- dist.res %>%
  mutate(paired = (pt.id.16s == pt.id.rna))
```

save distance table
```{r}
#write.csv(dist.res, "../data/16s_distances.csv")
dist.res <- read.csv("../data/16s_distances.csv")
```

relabel control 16s
```{r}
dist.res <- dist.res %>%
  mutate(paired = ifelse(is.na(paired), "CONTROL", paired))
```


plot 
```{r}
dist.res %>%
  mutate(paired.char = if_else(is.na(specimen.type),
                          true = "Control",
                          false = specimen.type)) %>%
  filter(paired != "CONTROL") %>% # remove control for exotic fig
  mutate(paired.char = as.character(paired.char)) %>%
  ggplot(aes(x=paired, y=value)) +
  geom_boxplot(aes(fill = paired.char)) +
  theme_bw(base_size = 7) +
  scale_fill_manual(values = c("salmon2", "cornflowerblue")) + 
  #scale_fill_viridis_d() +
  labs(x = "Paired Samples",
       y = "Bray-Curtis Distance from 16S to RNAseq",
       fill = "Tissue Source") +
  theme(legend.position = "top")
ggsave("../figures/16s_distance_boxplot.png", height=2.5, width=2.5)
```

```{r}
test.tum <- 
  dist.res %>%
  filter(specimen.type == "Tumor")
table(test.tum$paired)
```

```{r}
kw.tum.paired <- kruskal.test(value ~ paired, data = test.tum)
kw.tum.paired
```

The 16S samples are significantly more similar to RNAseq samples from the same patient than to different patients (Kruskal-Wallis rank sum test p-value = `r kw.tum.paired$p.value`).

```{r}
test.paired <- 
  dist.res %>%
  filter(paired == "TRUE")
table(test.paired$specimen.type)
```
```{r}
kw.spectype <- kruskal.test(value ~ specimen.type, data = test.paired)
kw.spectype
```

The RNAseq-derived microbes are not significantly different from 16S-derived microbes from tumors or adjacent normal tissue (Kruskal-Wallis rank sum test p-value = `r kw.spectype$p.value`).

```{r}
test.norm <- 
  dist.res %>%
  filter(specimen.type == "Normal")
table(test.norm$paired)
```

```{r}
kw.test.norm <- kruskal.test(value ~ paired, data = test.norm)
kw.test.norm
```