---
title: "correlate with age"
author: "Rebecca Hoyd"
date: "2/16/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(broom)
```

# Load data

```{r from drake freeze}
data.dir <- "/fs/ess/PAS1695/projects/exotic/data/drake-output/2022-03-08"
tcc.exo.ra <- readRDS(file.path(data.dir, "tcc.exora.taxonomy.RDS"))
tcga.exo.ra <- readRDS(file.path(data.dir, "tcga.exora.taxonomy.RDS"))

tcga.clin <- readRDS(file.path(data.dir, "tcga.clin.tum.RDS"))

tcc.cancer.labs <- readRDS(file.path(data.dir, "normalization.inputs.RDS"))[[2]] %>%
  filter(grepl("^SL", sample)) %>%
  select(sample,TCGA.code)

manifests <- list(read.csv("/fs/ess/PAS1695/generate-inputs_v2/TCGA-processing/data/COAD/manifest.csv"),
                  read.csv("/fs/ess/PAS1695/generate-inputs_v2/TCGA-processing/data/READ/manifest.csv")) %>%
  bind_rows() %>%
  mutate(sample.expression = file_id.expression) %>%
  select(sample.expression, file_name.expression)
```

```{r cibersort results}
ciber.files <- list.files("/fs/ess/PAS1695/projects/exotic/data/deconvolutions",
                          full.names = T)

ciber.tcc <- read.csv(ciber.files[1], stringsAsFactors = F) %>%
  mutate(sample = Input.Sample) %>%
  select(-X, -TILs, -cancer, -Input.Sample) 

ciber.tcga.good <- lapply(c(ciber.files[2], ciber.files[4:8]),
                          function(x) read.csv(x, stringsAsFactors = F)) %>%
  bind_rows() %>%
  mutate(sample.expression = coalesce(Input.Sample, Mixture)) %>%
  select(-Input.Sample, -P.value, -Pearson.Correlation, -RMSE, Mixture, -Correlation)

ciber.tcga.fix <- read.csv(ciber.files[3], stringsAsFactors = F) %>%
  mutate(file_name.expression = paste0(Mixture, ".htseq.counts.gz")) %>%
  left_join(manifests) %>%
  select(-Mixture, -Correlation, -RMSE, -P.value, -file_name.expression)

ciber.tcga <- bind_rows(ciber.tcga.good, ciber.tcga.fix)
```

# Functions

```{r}
exoRAtowide <- function(data, taxlev){
  tmp <- data
  tmp$Taxa <- tmp[[taxlev]]
  tmp.wide<- tmp %>%
    dplyr::filter(!is.na(Taxa)) %>%
    group_by(sample, Taxa) %>%
    summarise(ra = sum(exo.ra, na.rm = T)) %>%
    spread(key = "Taxa", value = "ra")
  
  tmp.wide[is.na(tmp.wide)] <- 0
  return(tmp.wide)
}

exoToDF <- function(taxalevels = c("domain", "kingdom", "phylum", "class", 
                                    "order", "family", "genus", "species"), 
                     data){
  w.ls <- lapply(taxalevels, function(x) exoRAtowide(data, x))
  w.df <- reduce(w.ls, function(x,y) left_join(x,y)) 
  return(w.df)
}
```

```{r}
correlate_mics_immune <- function(mics, immune, data, datalab){
  cor.res <- lapply(mics, function(m) lapply(immune, function(i)
    try(cor.test(data[[m]], data[[i]],
                 method = "spearman") %>%
          tidy() %>%
          mutate(ImmuneCell = i, microbe = m, datset = datalab))))
  
  
  test <- lapply(cor.res, function(x) bind_rows(x))
  cor.df <- bind_rows(test)
  return(cor.df)
}
```

# Formatting

```{r wide microbe data}
tcc.w <- exoToDF(data = tcc.exo.ra)
tcga.w <- exoToDF(data = tcga.exo.ra)
```

```{r combine with expression}
tcc.modin <- inner_join(ciber.tcc, tcc.w) %>%
  left_join(tcc.cancer.labs)

tcga.modin <- tcga.clin %>%
  select(file_id.BAM, file_id.expression, cancer) %>%
  mutate(sample = file_id.BAM, sample.expression = file_id.expression) %>%
  inner_join(tcga.w) %>%
  inner_join(ciber.tcga)
```

```{r grab gene and microbe lists}
microbes <- intersect(colnames(tcc.w[,-1]), colnames(tcga.w[,-1]))
imcells <- colnames(ciber.tcc)[1:20]

```

# Run correlations

```{r, warning=FALSE}
tcc.res <- correlate_mics_immune(microbes, imcells, tcc.modin, "ORIEN")
tcga.res <- correlate_mics_immune(microbes, imcells, tcga.modin, "TCGA")
```

```{r}
tcc.cancers <- unique(tcc.modin$TCGA.code)
tcga.cancers <- unique(tcga.modin$cancer)

tcc.cph.cancer <- lapply(tcc.cancers, function(y)
  correlate_mics_immune(microbes, imcells,
                      filter(tcc.modin,
                             TCGA.code == y),
                      y))

tcga.cph.cancer <- lapply(tcga.cancers, function(y) 
  correlate_mics_immune(microbes, imcells,
                      filter(tcga.modin,
                             cancer == y),
                      y))
```

```{r format split cancer results}
tcc.cancer.form <- bind_rows(tcc.cph.cancer) %>%
  mutate(cancer = datset,
         datset = "TCC") 
tcga.cancer.form <- bind_rows(tcga.cph.cancer) %>%
  mutate(cancer = datset,
         datset = "TCGA") 
```


# Save

```{r}
write.csv(tcc.res, "../data/correlations_mic-immune_TCC.csv", row.names = F)
write.csv(tcga.res, "../data/correlations_mic-immune_TCGA.csv", row.names = F)

write.csv(tcc.cancer.form, "../data/correlations_mic-immune_TCC_split-cancer.csv",
          row.names = F)
write.csv(tcga.cancer.form, "../data/correlations_mic-immune_TCGA_split-cancer.csv",
          row.names = F)
```