---
title: "Building the sample set"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(readxl)
library(tidyverse)
library(tableone)
library(flextable)
```

# Pull in all TCC data
```{r}
tcc.linkage <- read_excel("/fs/ess/PAS1695/exoticpipe/external-data/Additional Data_Spakowicz_SLids.xlsx")
tcc.meta <-read_excel("/fs/ess/PAS1695/exoticpipe/external-data/OSU.RNAseq.sequencing.QC.metrics.xlsx")

tcc.clin <- read.csv("/fs/ess/PAS1695/projects/exotic/data/2020-02-19_clinical_aggregated.csv", 
                     stringsAsFactors = F)
tcga.clin <- readRDS("/fs/ess/PAS1695/projects/exotic/data/drake-output/2022-03-08/tcga.clin.tum.RDS")
tcga.pass <- readRDS("/fs/ess/PAS1695/projects/exotic/data/drake-output/2022-03-08/tcga.counts.norm.RDS")
```

# Format to set we have

```{r}
combined.metdat <- tcc.clin %>%
  rename("tcc_id" = "RNA.SL.ID") %>%
  left_join(tcc.linkage)

combined.metdat.form <- combined.metdat %>%
  mutate(FFPE = `Tumor Tissue Sequenced Specimen Category`,
         tumor.location = `Tumor Tissue Sequenced Collection Site`,
         `BMI at Collection` = BMI.at.Collection,
         `Age at Collection` = Age.at.Collection,
         `Vital Status` = Vital.Status)
```

# Make table 1

```{r}
allvars <- c("BMI at Collection", "Age at Collection", "Gender", 
             # "Cancer.Stage.at.Collection",
             # "Chemotherapy", "I.O.Drug.Name", "Radiation", "Surgery", 
             "Vital Status", 
             "FFPE"
             # "tumor.location"
             )

catvars <- c("Gender", "Cancer.Stage.at.Collection", "Chemotherapy", "I.O.Drug.Name", "Radiation",
             "Surgery", "Vital Status", "FFPE", "tumor.location")

CreateTableOne(allvars, strata = "Cancer", data = combined.metdat.form)
  
```

```{r}
tab1 <- CreateTableOne(allvars, strata = "Cancer", data = combined.metdat.form)
tab1 <- print(tab1, printToggle = F)

tab1.flex <- tab1 %>%
  as.data.frame() %>%
  rownames_to_column(var = "Variables") %>%
  select(-test) %>%
  flextable()

tab1.flex
```

```{r}
save_as_docx(tab1.flex, path = "../tables/table1.docx")
```

# Check against TCGA

```{r}
tcc.met <- combined.metdat.form %>%
  mutate(metastatic = ifelse(grepl("4", Cancer.Stage.at.Collection), 1, 0),
         source = "ORIEN",
         sample = tcc_id) %>%
  select(sample, source, metastatic)

tcga.met <- tcga.clin %>%
  mutate(metastatic = ifelse(grepl("iv", tumor_stage), 1, 0),
         source = "TCGA",
         sample = file_id.BAM) %>%
  filter(sample %in% tcga.pass$sample) %>%
  select(sample, metastatic, source)


met.compare <- bind_rows(tcc.met, tcga.met)
```

```{r}
CreateTableOne("metastatic", "source", met.compare, "metastatic")
```
