---
title: "correlate with age"
author: "Rebecca Hoyd"
date: "2/16/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
```

# Load data

```{r}
data.dir <- "/fs/ess/PAS1695/projects/exotic/data/drake-output/2022-03-08"
tcc.exo.ra <- readRDS(file.path(data.dir, "tcc.exora.taxonomy.RDS"))
tcga.exo.ra <- readRDS(file.path(data.dir, "tcga.exora.taxonomy.RDS"))

tcc.expr <- readRDS(file.path(data.dir, "tcc.expr.norm.RDS"))
tcga.expr <- readRDS(file.path(data.dir, "tcga.expr.norm.RDS"))

tcga.clin <- readRDS(file.path(data.dir, "tcga.clin.tum.RDS"))
```

# Functions

```{r}

exoRAtowide <- function(data, taxlev){
  tmp <- data
  tmp$Taxa <- tmp[[taxlev]]
  tmp.wide<- tmp %>%
    dplyr::filter(!is.na(Taxa)) %>%
    group_by(sample, Taxa) %>%
    summarise(ra = sum(exo.ra, na.rm = T)) %>%
    spread(key = "Taxa", value = "ra")
  
  tmp.wide[is.na(tmp.wide)] <- 0
  return(tmp.wide)
}

exoToDF <- function(taxalevels = c("domain", "kingdom", "phylum", "class", 
                                    "order", "family", "genus", "species"), 
                     data){
  w.ls <- lapply(taxalevels, function(x) exoRAtowide(data, x))
  w.df <- reduce(w.ls, function(x,y) left_join(x,y)) 
  return(w.df)
}
```

```{r}
correlate_mics_genes <- function(mics, genes, data, datalab){
  cor.res <- lapply(mics, function(m) lapply(genes, function(i)
    try(cor.test(data[[m]], data[[i]],
                 method = "spearman") %>%
          tidy() %>%
          mutate(Gene = i, microbe = m, datset = datalab))))
  
  
  test <- lapply(cor.res, function(x) bind_rows(x))
  cor.df <- bind_rows(test)
  return(cor.df)
}
```

# Formatting

```{r wide microbe data}
tcc.w <- exoToDF(data = tcc.exo.ra)
tcga.w <- exoToDF(data = tcga.exo.ra)
```

```{r combine with expression}
tcc.modin <- inner_join(tcc.expr, tcc.w)

tcga.modin <- tcga.clin %>%
  select(file_id.BAM, file_id.expression) %>%
  mutate(sample = file_id.BAM) %>%
  inner_join(tcga.w) %>%
  mutate(sample = file_id.expression) %>%
  inner_join(tcga.expr)
```

```{r grab gene and microbe lists}
mics <- intersect(colnames(tcc.w[,-1]), colnames(tcga.w[,-1]))
genes <- intersect(colnames(tcc.expr[,-1]), colnames(tcga.expr[,-1]))

genes.ls <- split(genes, floor(seq_along(genes)/100))
```

```{r}
saveRDS(tcc.modin, "/fs/ess/PAS1695/projects/exotic/data/mic-genes_TCC.RDS")
saveRDS(tcga.modin, "/fs/ess/PAS1695/projects/exotic/data/mic-genes_TCGA.RDS")

saveRDS(mics, "/fs/ess/PAS1695/projects/exotic/data/microbe-names.RDS")
saveRDS(genes.ls, "/fs/ess/PAS1695/projects/exotic/data/gene-lists.RDS")
```

# Run correlations

```{r}
  fileOut<- paste0("/fs/ess/PAS1695/projects/exotic/scripts/batch/corr-mic-genes.pbs")
  
  writeLines(c(paste0("#PBS -N corr-mic-genes_", f),
               "#PBS -A PAS1695",
               "#PBS -l walltime=12:00:00",
               "#PBS -l nodes=1:ppn=28",
               "#PBS -j ae",
               "",
               "module load R/4.1.0-gnu9.1",
               paste0("Rscript ~/Documents/repos/exoTCC/manuscripts/exoTiC/",
                      "scripts/corr_functions.R $PBS_ARRAY_INDEX")),
             fileOut
  )
  # close(fileOut)

```

# Display, save
