---
title: "Correlations with taxa, TILs and gene expression"
author: "Daniel Spakowicz"
date: "12/9/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(dplyr)
library(RColorBrewer)
library(ggplot2)
library(forcats)
library(broom)
library(tidyr)
```

```{r exogenous sequences relative abundances}
exo.tax <- read.csv("../data/agg_exo-ra_taxonomy.csv", stringsAsFactors = FALSE)
key <- read.csv("../data/exotcc_deidentified-ids.csv", stringsAsFactors = FALSE) %>%
  rename("sample" = "id")
imcells <- read.csv("../data/split-deconvolved-aggcuff_absolute/2019-12-09_cibersort_absolute_aggregated+TILS.csv",
                    stringsAsFactors = FALSE)
```


# Stacked bar at Phylum level

## All data
```{r}
exo.gi <- exo.tax %>%
  left_join(key) %>%
  filter(cancer == "GI")

badspecs <- exo.gi %>%
  filter(is.na(taxonomy)) %>%
  mutate(domain = ifelse(grepl("phage|virus", spec.join, ignore.case = TRUE), "Viruses", domain),
         domain = ifelse(grepl("calothrix", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("anabaena", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("chondrocystis", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("cyanobacterium", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("dolichospermum", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("euhalothece", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("fischerella", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("geitlerinema", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("geminocystis", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("gloeobacter", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("gloeothece", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("limnospira", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("microcoleus", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("moorea", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("nostoc", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("oscillatoria", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("prochlorococcus", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("rivularia", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("synechococcus", spec.join, ignore.case = TRUE) & !grepl("phage|virus", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("tenericutes", spec.join, ignore.case = TRUE), "Bacteria", domain),
         domain = ifelse(grepl("trichodesmium", spec.join, ignore.case = TRUE), "Bacteria", domain)
         )

exo.gi <- exo.gi %>%
  filter(!(spec.join %in% badspecs$spec.join))

exo.gi <- bind_rows(exo.gi, badspecs)
```

```{r}
exo.gi.dom <- exo.gi %>%
  mutate(domain = ifelse(spec.join == "Proteus.phage.VB_PmiS.Isfahan", "Viruses", domain)) %>%
  group_by(sample, domain) %>%
  summarise(domain.ra = sum(ra, na.rm = TRUE))

doms <- unique(exo.gi.dom$domain)
dom.count <- length(doms)
getPal <- colorRampPalette(brewer.pal(8, "Dark2"))

getord <- exo.gi.dom %>%
  group_by(domain) %>%
  summarise(Mean = mean(domain.ra, na.rm = TRUE)) %>%
  arrange(Mean)
```



```{r}
exo.gi.dom %>%
  ggplot(aes(x = sample, y = domain.ra, fill = fct_relevel(domain, rev(getord$domain)))) +
  geom_col() +
  scale_fill_manual(breaks = doms, values = getPal(dom.count), na.value = "grey", name = "Domain") +
  labs(x = "", y = "Relative Abundance") +
  theme_bw() +
  theme(axis.ticks.x = element_blank(), axis.text.x = element_blank()) +
  ggsave("../figures/gi_stacked-bar.png")
```



# Heatmap of phyla correlations to TIL's

## Calculate correlations

```{r}
exo.gi.phyl <- exo.gi %>%
  mutate(phylum = ifelse(spec.join == "Proteus.phage.VB_PmiS.Isfahan", "Proteus phage", phylum)) %>%
  filter(!is.na(phylum)) %>%
  group_by(sample, phylum) %>%
  summarise(phylum.ra = sum(ra, na.rm = TRUE))

phyls <- unique(exo.gi.phyl$phylum)
phyl.count <- length(phyls)
getPal <- colorRampPalette(brewer.pal(8, "Dark2"))

getord <- exo.gi.phyl %>%
  group_by(phylum) %>%
  summarise(Mean = mean(phylum.ra, na.rm = TRUE)) %>%
  arrange(desc(Mean))

phyls.abun <- getord$phylum[1:15]
```

```{r}
int.cells <- c("T.cells.CD8", "NK.cells.resting", "NK.cells.activated", "T.cells.regulatory..Tregs.", "Neutrophils", "TILs")
# phyls.abun <- ifelse(is.na(phyls.abun), "<NA>", phyls.abun)
imcells.gi <- imcells %>%
  filter(cancer == "GI") %>%
  select(Input.Sample, int.cells)

exo.gi.imphyl <- exo.gi.phyl %>%
  spread(key = "phylum", value = "phylum.ra") %>%
  rename("Input.Sample" = "sample") %>%
  left_join(imcells.gi)
```

```{r}
cor.res <- list()
cor.res.cell <- list()
for(p in phyls.abun){
  for(i in int.cells){
    cor.res.cell[[i]] <- cor.test(as.numeric(exo.gi.imphyl[[p]]), as.numeric(exo.gi.imphyl[[i]]), 
                                  method = "spearman") %>%
      tidy() %>%
      select(estimate, p.value) %>%
      mutate(cell = i, phylum = p)
  }
  cor.res[[p]] <- cor.res.cell
}
```

```{r}
cor.df <- bind_rows(lapply(cor.res, function(x) bind_rows(x)))
```

## Build heatmap

```{r}
cor.df %>%
  mutate(dir = ifelse(estimate < 0, "neg", "pos")) %>%
  ggplot(aes(x = cell, y = fct_relevel(phylum, rev(phyls.abun)), fill = p.value, shape = dir)) +
  geom_tile() +
  geom_point(size = 3) +
  scale_fill_distiller(guide = "legend", breaks = c(0.00001, 0.01, 0.02, 0.03, 0.04, 0.05, 0.1, 0.2, 0.5, 1), 
                       limits = c(0, 1), values = c(0.00001, 0.01, 0.02, 0.03, 0.04, 0.05, 0.1, 0.2, 0.5, 1),
                       type = "div", palette = "BrBG", direction = 1, 
                       labels = c("<0.01", "0.01", "0.02","0.03", "0.04","0.05","0.1", "0.2", "0.5", "1"),
                       name = "P.value") +
  scale_shape_manual(breaks = c("neg", "pos"), labels = c("Negative", "Positive"), values = c("-", "+"),
                     name = "Correlation direction") +
  scale_x_discrete(breaks = int.cells, labels = c("CD8 T-cells", "Resting NK cells", "Activated NK cells",
                                                  "Regulatory T cells", "Neutrophils", "TIL")) +
  labs(x = "", y = "") +
  theme_bw() +
  theme(axis.text.x = element_text(hjust = 1, angle = 45)) +
  ggsave("../figures/gi_corr-pvalue-heatmap.png", width = 5, height = 7)
```



