---
title: "lit_processing"
author: "Mitchell Muniak"
date: "3/10/2021"
output: html_document
---

```{r setup, include=FALSE}
library(readxl)
library(dplyr)
source("00-paths.R")
```

# Processing
+ My goal is to create a data frame with three columns: Reference, Taxon, and Prevalence

### Load data frame
```{r}
# Load in table from my literature search
lit_data <- read_excel(file.path(paths$box.exogieo, "taxa_table_r_compatible_taxa_filled.xlsx")) %>%
  # Remove unneeded columns
  select(-sample_size, -method, -"patient/sample info", -relative_abundance, -classification, -organism) %>%
  # Remove rows that lack prevalence values
  # Remove the Yongzhi Yang reference. We are using Yang's supplementary data instead. This data is processed in "y-yang_prevalence.RMD"
  filter(!is.na(Prevalence), Ref != "Yongzhi Yang") %>%
  # A slight rename to the reference names to make them machine readable. This becomes relevant when we are saving our data frames in this script.
  mutate(Ref = gsub(" ", "_", Ref),
         Reference = gsub("Aleksandar_D._Kostic", "Kostic, et al. Genome Research (2012)", Ref),
         Reference = gsub("Mauro_Castellarin", "Castellarin, et al. Genome Research (2012)", Reference),
         Reference = gsub("Yannick_Eisele", "Yannick, et al. Clinical Colorectal Cancer (2021)", Reference),
         Reference = gsub("Shuping_Long", "Long, et al. npj Biofilms and Microbiomes (2020)", Reference),
         Reference = gsub("Jinho_Yang", "Yang J, et al. Experimental & Molecular Medicine (2019)", Reference))
```

### Format literature taxonomy
```{r}
# Add suffix to match each taxonomic level
lit_data$domain <- paste("d__", lit_data$domain, sep = "")
lit_data$phylum <- paste("p__", lit_data$phylum, sep = "")
lit_data$class <- paste("c__", lit_data$class, sep = "")
lit_data$order <- paste("o__", lit_data$order, sep = "")
lit_data$family <- paste("f__", lit_data$family, sep = "")
lit_data$genus <- paste("g__", lit_data$genus, sep = "")
lit_data$species <- paste("s__", lit_data$species, sep = "")
```

```{r}
# Revert NAs
# The above block of code added a suffix to NA values, we want to return these values to NA
lit_data$domain <- na_if(lit_data$domain, "d__NA")
lit_data$phylum <- na_if(lit_data$phylum, "p__NA")
lit_data$class <- na_if(lit_data$class, "c__NA")
lit_data$order <- na_if(lit_data$order, "o__NA")
lit_data$family <- na_if(lit_data$family, "f__NA")
lit_data$genus <- na_if(lit_data$genus, "g__NA")
lit_data$species <- na_if(lit_data$species, "s__NA")
```

```{r}
# We want to consolidate all the taxons from the lit search into a single "taxon" column
# Unlike with our data, we do not want to aggregate this data. 
# We only want to use the most specific taxonomic level for each organism.
prev_lit <- lit_data %>%
  # Nesting several if_else statements allowed for selection of the highest taxonomic level
  mutate(Taxon = if_else(is.na(genus),
            true = if_else(is.na(family),
                           true = if_else(is.na(order),
                                          true = if_else(is.na(class),
                                                         true = if_else(is.na(phylum),
                                                                        true = domain,
                                                                        false = phylum),
                                                         false = class),
                                        false = order), 
                           false = family),
            false = genus)) %>%
  select(Ref, Taxon, Prevalence, Reference) 
```

```{r}
# Split data frame into list based on reference variable
prev_lit_split <- split(prev_lit, prev_lit$Ref)
```

```{r}
# The purpose of this chunk is to show what data frames are created in this script and what they are named.
# Please view the results of this chunk to see the names of these data frames
paste0(unique(prev_lit$Ref), "_prev.csv")
```

### Save
```{r, results='hide'}
# Save each new data frame, with each data frame named after the reference variable it contains
lapply(prev_lit_split, function(prev_lit){
  write.csv(prev_lit, paste0("../data/", unique(prev_lit$Ref), "_prev.csv"), row.names = FALSE)
})
```